core/src/main/java/io/questdb/log/LogRecordSink.java

```diff
@@ -59,28 +63,121 @@ public int length() {
 
     @Override
     public CharSink put(CharSequence cs) {
-        int rem = (int) (lim - _wptr);
-        int len = cs.length();
-        int n = Math.min(rem, len);
-        Chars.asciiStrCpy(cs, n, _wptr);
-        _wptr += n;
+        final int rem = (int) (lim - _wptr);
+        final int len = cs.length();
+        if (rem >= len) {
+            // Common case where the buffer fits the available space.
+            Chars.asciiStrCpy(cs, len, _wptr);
+            _wptr += len;
+            return this;
+        }
+
+        // The line is being truncated:
+        // We determine a safe length to byte-copy.
+        // We skip copying the last 4 bytes, as they may be a multi-byte UTF-8 character.
+        // NOTE: The computed length may be negative.
+        int safeLen = rem - 4;
```

Comment:
I guess there was confusion as to what this method is used for.

Context: this is logger and not a generic one, logger for QuestDB. All our fixed strings are English ASCII. E.g. There are no deliberate strings constants that are non-ascii.

There are two cases where we may have anything non-ascii:
1. UTF16 encoded CharSequence, in which case code must explicitly call `log.utf8(CharSequence sequence)`
2. Somethings on the wire in UTF8 format, which we would like to log correctly. In this case we call your function `log.utf8(lo,hi)`

Therefore assumptions about multi-byte characters here will not hold true

Reply:
Rolled back changes for `put(CharSequence, ..)` after your clarification on their use case (and some digging to confirm it).